sstables/processing_result_generator.hh

```diff
@@ -0,0 +1,127 @@
+ * To execute a fragment of the coroutine that ends at an co_yield, and get the yielded value use generate().
+ * Each subsequent generate() call starts after the last co_yield and continues until another one is encountered.
+ *
+ * Explanation of C++ coroutines below:
+ * A "co_yield val" call is equivalent to "co_await yield_value(val)". In yield_value we save the value in
+ * "current_value", and it returns a suspend_always object, causing the coroutine to return to the caller after
+ * each co_yield.
+ * The await_transform() method is deleted, so we don't use co_await directly without yielding any value.
+ * The unhandled_exception() method is called when an unhandled exception is detected in the coroutine. We
+ * save it to a local variable, and check it in our generate() calls.
+ */
+class processing_result_generator {
+public:
+    struct promise_type;
+    struct read_awaiter {
+        data_consumer::read_status _rs;
+        read_awaiter(data_consumer::read_status rs) : _rs(rs) {}
+        constexpr bool await_ready() const noexcept { return _rs == data_consumer::read_status::ready; }
+        constexpr void await_suspend(std::experimental::coroutine_handle<promise_type>) const noexcept {}
+        constexpr void await_resume() const noexcept {}
+    };
+    struct promise_type {
+        using handle_type = std::experimental::coroutine_handle<promise_type>;
+        processing_result_generator get_return_object() {
+            return processing_result_generator{handle_type::from_promise(*this)};
+        }
+        // the coroutine doesn't start running until the first handle::resume() call
+        static std::experimental::suspend_always initial_suspend() noexcept {
+            return {};
+        }
+        static std::experimental::suspend_always final_suspend() noexcept {
+            return {};
+        }
+        std::experimental::suspend_always yield_value(data_consumer::processing_result value) noexcept {
+            current_value = std::move(value);
+            return {};
+        }
+        read_awaiter yield_value(data_consumer::read_status rs) noexcept {
+            if (rs == data_consumer::read_status::waiting) {
+                current_value = data_consumer::proceed::yes;
+            }
+            return read_awaiter(rs);
+        }
+        // Disallow co_await in generator coroutines.
+        void await_transform() = delete;
+
+        void unhandled_exception() {
+            caught_exception = std::current_exception();
+        }
+        void return_void() noexcept {}
+
+        data_consumer::processing_result current_value;
+        std::exception_ptr caught_exception;
+    };
+private:
+    std::experimental::coroutine_handle<promise_type> m_coroutine;
+public:
+    explicit processing_result_generator(const std::experimental::coroutine_handle<promise_type> coroutine) :
+        m_coroutine{coroutine}
+    {}
+
+    processing_result_generator() = default;
```

Comment:
We probably don't want to allow creating a generator with no coroutine (so we won't accidentaly try to generate data from it), so I'll change this to delete